streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedTasks.java

```diff
@@ -103,6 +103,7 @@ void addNewTask(final T task) {
                     log.debug("Transitioning {} {} to restoring", taskTypeName, entry.getKey());
                     addToRestoring(entry.getValue());
                 } else {
+                    entry.getValue().initializeTopology();
```

Comment:
In the above line 102 calling `initialize()`, both `initializeStateStores()` which registers the stores for restoration and the `initTopology()`. And even if it returns false these functions are already called, so I'm not sure if it will work..

I'd suggest doing the following: 

1. having the `AssignedTasks<T extends AbstractTask>`.
2. let `initializeStateStores` returns a boolean just like `initialize`, then in the line 103 above, call `AbstractTask#initializeStateStores` instead.
3. remove `initializeStateStores` from `initialize` of stream task and standby task.

Reply:
I get your point, but I'm a little confused from your comment as in `initialize()` I've moved the `initToplogy` and `ProcessorContext` initialization to the new `Task.initializeTopology` method. But having done that I agree that `initialize` is not properly named and should be renamed to `initializeStateStores`.

Since we are breaking the initialization into two phases, IMHO we need two methods on the `Task` interface, `intitializeStateStores` and `initializeTopology` and keep `AssignedTasks<T extends Task>`.

WDYT?
\cc @mjsax